Add script to help "unstick" military dwarves' uniform equiping.

roccomoretti opened the pull request:
One of the long-standing issues with military dwarves is that they can often get "stuck" when putting on their uniform, such that they never actually equip their uniform items. This script is an attempt to address that, in a minimally "cheaty" way.

One cause is of the issue is that dwarves are not smart enough to remove a non-uniform piece from a body part to put on their uniform piece. This script can cause dwarves to drop onto the floor all items on a body part which is missing a uniform piece. Another cause is when the assigned item is held by someone or something else, so the script can also force the assigned items to be removed from whatever inventory/container they're currently in and be placed on the floor. The script does not attempt to assign an item if one isn't already (i.e. if the green check is missing from the military equip screen), and only prints info by default.

This script can achieve much the same thing as the "replace clothing" designation, but with the twin benefits that a) it will only cause disrobing for body parts which are missing uniform items (and not ones where uniform and clothing peacefully coexist) and b) it only functions where there is an assigned uniform item, and doesn't force your dwarves to go barefoot until the armorer finishes making high boots.

In my limited testing, the script is successful in prompting dwarves to pick up the needed armor, though it can take a while and several applications if the dwarf is missing many items.

The script is provisionally called "uniforce" (uniform + force), but that reflected mostly the initial rough concept, and may not be the best name for the current approach. Suggestions for a more appropriate name are welcome.

This is my first significant attempt at both Lua scripting and working with DFHack, so please let me know if there's anything I should be doing differently.

myk002 commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+
+-- Will figure out which items need to be moved to the floor, returns an item_id:item map
+function process(unit, args)
+  silent = args.all -- Don't print details if we're iterating through all dwarves
```
missing 'local'?

myk002 commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+    covered = {} -- Don't consider current covers - drop for anything which is missing
+  end
+
+  -- Figure out body parts which should be covered but aren'y
```
aren'y

myk002 commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+  for id, item in pairs(item_list) do
+    local pos = get_item_pos(item)
+    if pos == nil then
+      dfhack.println("Could not find drop location for item #"..id.."  "..utils.getItemDescription(item))
```
dfhack.println is aliased to 'print', which is shorter and cleaner

lethosor replied:
`dfhack.printerr` could also be an alternative for messages that are errors (it does not have an alias, though). I'm guessing `qerror` would not be appropriate since you don't want to actually raise an error here (since you're processing a list).

myk002 commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+end
+
+if (args.drop or args.free) and df.global.ui.main.mode == df.ui_sidebar_mode.ViewUnits then
+  dfhack.println("Error: Cannot actually drop/free items when view-unit sidebar is open.")
```
you can set the mode and update the sidebar instead of failing here. just restore the original mode before returning. see https://github.com/myk002/scripts/blob/master/internal/quickfort/query.lua#L100 for an example

lethosor replied:
Also: convention for fatal errors is `qerror`, which will throw an error without a traceback (and terminate the script)

myk002 commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+Options:
+:(none):      Simply show identified issues (dry-run).
+:-drop:       Cause offending worn items to be placed on ground under unit.
+:-free:       Remove to-equip items from containers or other's inventories, and place on ground.
```
does this interact cleanly with items that are being hauled?

roccomoretti replied:
It uses `dfhack.items.moveToGround()`, so it should work as cleanly as that does. (And if, for some reason, that function doesn't work cleanly, then I'd probably argue the fix should be in that function.)

lethosor commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+    end
+  end
+else
+  local unit=dfhack.gui.getSelectedUnit(true)
+  if df.isvalid(unit) then
+    local to_drop = process(unit,args)
+    do_drop( to_drop )
+  else
+    dfhack.println("No unit is selected")
+  end
```
You could also use qerror here, but my suggestion would be to remove the argument to getSelectedUnit, which will cause it to print an error message, then only continue if a unit is set:
```suggestion
  local unit = dfhack.gui.getSelectedUnit()
  if unit then
    local to_drop = process(unit,args)
    do_drop( to_drop )
  end
```

lethosor commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+    dfhack.println("NOT VISIBLE TILE")
+    return nil
+  end
+  return {x=x, y=y, z=z}
```
There's a built-in function for this: `xyz2pos(x, y, z)`

lethosor commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+  return nil
+end
+
+-- This should be okay for dwarves, not sure if valid for others.
```
Probably unlikely that it'll work for _all_ other races, although humans might be similar enough. It would be great if this could be looked up somehow, but I'm no expert on this. It appears that the caste raws have quite a bit of detail about body parts/layers/etc, accessible through e.g. `df.creature_raw.find(unit.race).caste[unit.caste].body_info` - maybe some of that is relevant. For instance, `body_info.body_parts[15]` of an unmodded dwarf has `category = "FOOT"` and `flags.STANCE = true` (as well as two unidentified flags set, 28 and 29).

lethosor reviewed (commented):
Cool! Thanks for handling documentation, and fixing the linter stuff.
Regarding naming, I'm also not sure what a good alternative to `uniforce` is. I thought of `reset-uniform`, but that might sound like `clear-uniform` to some people, which isn't what this is doing. We want it to be discoverable from the script listing, of course, so I think having the word `uniform` in there would be good (even though it's almost there, some people could latch onto the `force` part, since there's an existing `force` script). Naming is hard...

AtomicChicken commented on uniforce.lua:
```diff
@@ -0,0 +1,259 @@
+  return {x=x, y=y, z=z}
+end
+
+function find_item_by_ID(id)
```
This function can be replaced with `df.item.find(id)`.

roccomoretti commented:
I've updated the script based on the suggestions (thank you, all). The biggest change is probably the reworking to use the `bodyparts_that_can_wear(unit, item)` function rather than the uniform armor location designation.

I've also renamed the script `uniform-unstick`. Again, not necessarily the best name, but hopefully slightly better than it was.

lethosor commented on uniform-unstick.lua:
```diff
@@ -0,0 +1,277 @@
+  if item._type == df.item_helmst then
+    for index, part in pairs(unitparts) do
+      if part.flags.HEAD then
+        bodyparts[#bodyparts+1] = index
```
`table.insert` is a built-in function that does what you want here (and below): 
```lua
        table.insert(bodyparts, index)
```

lethosor reviewed (commented):
Thanks for tackling this! `uniform-unstick` sounds good to me.

lethosor reviewed (approved):
Looks good, and thanks! I fixed a bug I spotted in a fort - hopefully that was all.

I also made a couple documentation tweaks - specifically, the lists starting with `:` (ReStructuredText calls them "definition lists") needed to have a blank line before them, to keep them from looking like this:
![image](https://user-images.githubusercontent.com/3719547/95414400-d0555980-08fb-11eb-8e9b-98a9399d2745.png)

I wish there were a more obvious warning/error for things like these, but as-is, you'd have to build the documentation to catch them. We have a page [here](https://docs.dfhack.org/en/latest/docs/Documentation.html) that goes into some detail about how to do that, if you're interested, although it's a bit involved.
